math/matrix/src/TMatrixTSparse.cxx

```diff
@@ -491,111 +543,119 @@ void TMatrixTSparse<Element>::ExtractRow(Int_t rown, Int_t coln, Element *v,Int_
-         if (pRowIndexa[irowa] >= pRowIndexa[irowa+1]) continue;
-         for (Int_t irowb = 0; irowb < b.GetNrows(); irowb++) {
-            if (pRowIndexb[irowb] >= pRowIndexb[irowb+1]) continue;
-            pRowIndexc[irowa+1]++;
-            pColIndexc[ielem++] = irowb;
-         }
-      }
-   } else {
-      pRowIndexc = this->GetRowIndexArray();
-      pColIndexc = this->GetColIndexArray();
+      if (nc == 0)
+         return;
    }
 
-   const Element * const pDataa = a.GetMatrixArray();
-   const Element * const pDatab = b.GetMatrixArray();
-   Element * const pDatac = this->GetMatrixArray();
-   Int_t indexc_r = 0;
-   for (Int_t irowc = 0; irowc < this->GetNrows(); irowc++) {
-      const Int_t sIndexa = pRowIndexa[irowc];
-      const Int_t eIndexa = pRowIndexa[irowc+1];
-      for (Int_t icolc = 0; icolc < this->GetNcols(); icolc++) {
-         const Int_t sIndexb = pRowIndexb[icolc];
-         const Int_t eIndexb = pRowIndexb[icolc+1];
-         Element sum = 0.0;
-         Int_t indexb = sIndexb;
-         for (Int_t indexa = sIndexa; indexa < eIndexa && indexb < eIndexb; indexa++) {
-            const Int_t icola = pColIndexa[indexa];
-            while (indexb < eIndexb && pColIndexb[indexb] <= icola) {
-               if (icola == pColIndexb[indexb]) {
-                 sum += pDataa[indexa]*pDatab[indexb];
-                 break;
-               }
-               indexb++;
-            }
+   Int_t *pRowIndex = this->GetRowIndexArray();
+   Int_t *pColIndex = this->GetColIndexArray();
+   Element *pData = this->GetMatrixArray();
+
+   // we compute each column of the result, one after the other
+   for (Int_t j = 0; j < cols; ++j) {
+      Int_t nnz = 0;
+      for (Int_t l = pRowIndexrhs[j]; l < pRowIndexrhs[j + 1]; ++l) {
+         double y = *(rhsVal + l);
+         Int_t k = *(rhsCol + l);
+         for (Int_t m = pRowIndexlhs[k]; m < pRowIndexlhs[k + 1]; ++m) {
+            Int_t i = *(lhsCol + m);
+            double x = *(lhsVal + m);
+            if (!mask[i]) {
+               mask[i] = true;
+               values[i] = x * y;
+               indices[nnz] = i;
+               ++nnz;
+            } else
+               values[i] += x * y;
          }
-         if (sum != 0.0) {
-            pColIndexc[indexc_r] = icolc;
-            pDatac[indexc_r] = sum;
-            indexc_r++;
+      }
+      pRowIndex[j + 1] = pRowIndex[j];
+      for (Int_t i = 0; i < rows; ++i) {
+         if (mask[i]) {
+            mask[i] = false;
+            Int_t p = pRowIndex[j + 1];
+            ++pRowIndex[j + 1];
+            pColIndex[p] = i;
+            pData[p] = values[i];
          }
       }
-      pRowIndexc[irowc+1] = indexc_r;
    }
 
-   if (constr)
-      SetSparseIndex(indexc_r);
+   if (gMatrixCheck) {
+      if (this->fNelems != pRowIndex[cols]) {
+         Error("conservative_sparse_sparse_product_impl", "non zeros numbers do not match");
+         return;
+      }
+   }
+
+   delete[] mask;
```

Comment:
Aren't we leaking these arrays when early returning? Maybe it would be best to use unique_ptrs instead?

Reply:
Thanks for pointing this out, should be fixed now!